src/shims/foreign_items.rs

```diff
@@ -13,6 +13,36 @@ use crate::*;
 use super::backtrace::EvalContextExt as _;
 use helpers::{check_abi, check_arg_count};
 
+/// This macro behaves just like `match $link_name { ... }`, but inserts a
+/// `$crate::helpers::check_abi($abi, $exp_abi)?` call at each match arm
+/// except the wildcard one.
+#[macro_export]
+macro_rules! match_with_abi_check {
+    ($link_name:expr, $abi:expr, $exp_abi:expr, {
+        $(|)? $($pattern:pat)|+ $(if $guard:expr)? => $shim_impl:block
+        $($remaining:tt)+
+    }) => {
+        match ($link_name, $abi, $exp_abi) {
+            ($($pattern)|+, abi, exp_abi) $(if $guard)? => {
+                $crate::helpers::check_abi(abi, exp_abi)?;
+                $shim_impl
+            }
+            (link_name, abi, exp_abi) => match_with_abi_check!(
+                link_name,
+                abi,
+                exp_abi,
+                { $($remaining)* }
+            ),
+        }
```

Comment:
So... this creates as many nested matches as there are patterns? Can this macro be rewritten in a way that it expands all arms next to each other?

Reply:
I'm honestly not convinced that this macro is a good idea. I think I'd prefer just having an ABI check in each arm manually. That's not enough overhead to justify the complexity of such a macro, IMO.

Reply:
That's reasonable... I used a regex find-and-replace to insert them... Hopefully that doesn't make a mess.